CL-1092 Implement ref distribution for birth-year custom field

adessy opened the pull request:
See jira ticket for description: CL-1092 (User story: CL-1273)

---


### TODO 

- [ ] Remove probabilities from CategoricalDistribution? (ask luuc if he's using them)

adessy committed:
CL-1092 Add type column to ref_distributions table

The `type` column will be used to implement multiple types of reference distributions using STI.

adessy committed:
CL-1092 Split `RefDistribution` into `RefDistribution` and `CategoricalDistribution`

`RefDistribution` is now the base class for the STI implementation of different types of distributions.

adessy committed:
CL-1092 Fix an incorrect method name

adessy committed:
CL-1092 Make some RefDistribution attributes read-only

adessy committed:
CL-1092 Fix rubocop offenses

adessy committed:
CL-1092 Rename :ref_distribution factory to :categorical_distribution

adessy committed:
CL-1092 Implement basic BinnedDistribution

Implement a first version of BinnedDistribution. Binned distribution are reference distribution for birth-year custom fields.

adessy committed:
CL-1092 Enable the creation of reference distributions for birthyear custom field

adessy committed:
CL-1092 Validate the type attribute of reference distributions

adessy committed:
CL-1092 Validate distributions with json schema

adessy committed:
CL-1092 Merge branch 'master' into CL-1092/binned-counts-reference-distribution

# Conflicts:
#	back/db/schema.rb

adessy committed:
CL-1092 Update schema.rb

Re-run migration after merge from master.

adessy committed:
CL-1092 Add factory for binned distributions

adessy committed:
CL-1092 Add tests for binned distributions

adessy committed:
CL-1092 Rework reference distribution tests

Extracts the tests specific to categorical distributions from ref_distribution_spec.rb to categorical_distribution_spec.rb. The remaining tests that are applicable to all types of distributions are used as shared examples that are included in the test files for reference distributions.

adessy committed:
CL-1092 Fix rubocop offenses

adessy committed:
CL-1092 Fix tests

It seems like `activerecord_json_validator` needs this method public.

adessy committed:
CL-1092 Rework acceptance tests for reference distributions

Create groups of shared examples for:
- unauthorized requests (visitors and normal users)
- reference distribution endpoints whose behavior does not depend on the type of distribution

Tests specific to categorical distributions have been moved to categorical_distributions_spec.rb.

adessy committed:
CL-1092 Allow GET and DELETE for binned distribution

Adapt the controller for reference distributions to support GET and DELETE for binned distribution.

Include 'reference distribution API' shared examples to acceptance tests for binned distribution to test those modifications.

adessy committed:
CL-1092 Rework the inference of the custom field

Give the path a higher priority than the query parameter when inferring the custom field.

adessy committed:
CL-1092 Add users_by_age stats endpoint

To calculate the age distribution, the years of birth must be converted to age and grouped into bins (default bins or bins defined by the reference distribution linked to the birthyear custom field, if any).

adessy committed:
CL-1092 Remove dead code

adessy committed:
CL-1092 Extracting age-stats logic from the controller

Extracting calculation and serialization of age stats from the controller. I also renamed some fields/attributes along the way.

adessy committed:
CL-1092 Add `users_by_age_as_xlsx` endpoint

adessy committed:
CL-1092 Fix `users_by_birthyear...` endpoints

Fix `users_by_birthyear...` endpoints and add regressions tests.

adessy committed:
CL-1092 Add basic tests for distribution serializers

Due to the polymorphic nature of reference distributions, acceptance tests for reference distributions do not longer test the exact responses (since the response is different for each distribution type). Tests just verify that the response matches the output of the serializer. Thus, I'm adding some tests for serializers to cover the exact payloads.

adessy committed:
CL-1092 Fix method name

adessy committed:
CL-1092 Add tests for UserCustomFields::StatUserPolicy

adessy committed:
CL-1092 Ensure bin boundaries are distinct for binned distributions

adessy committed:
CL-1092 Merge branch 'master' into CL-1092/binned-counts-reference-distribution

adessy commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb:
```diff
@@ -89,23 +102,31 @@ def find_users
+        end
+
+        def calculate_expected_user_counts
+          return if custom_field.key == 'birthyear'
```
I just added a new guard clause here.

adessy commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb:
```diff
@@ -89,23 +102,31 @@ def find_users
+        end
+
+        def calculate_reference_population
+          return if custom_field.key == 'birthyear'
```
I just added a new guard clause here.

adessy committed:
CL-1092 Improve documentation

adessy committed:
CL-1092 Remove unnecessary rounding

adessy committed:
CL-1092 Protect binned distribution from unwanted modifications

We freeze +BinnedDistribution#counts+ and +BinnedDistribution#bin_boundaries+ to prevent in-place modifications that would affect the reference distribution.

adessy commented:
Hey @koendehondt 👋 James being on leave, would you mind reviewing this one for me, plz? 

I've been through quite a few interruptions while working on it, so I have a hunch that I might have overlooked a few things. On the other hand, I need to have this rolled out next week at the latest (before my vacation). So all in all, I hope we'll find a way to make happen by then.

It's a pretty big chunk. So if you need some help to get the big picture, I'm down for a tuple session 🙂

adessy committed:
CL-1019 Reduce code duplication with shared examples

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -3,14 +3,19 @@
 module UserCustomFields
   module WebApi::V1
     class RefDistributionsController < ::ApplicationController
+      # Shortcuts for long namespace
+      RefDistribution = Representativeness::RefDistribution
+      CategoricalDistribution = Representativeness::CategoricalDistribution
+      private_constant :RefDistribution, :CategoricalDistribution
+
```
Wouldn't it be possible to move the constants to the `private` section, so that using `private_constant` is not necessary?

adessy replied:
I think the private section does not change the visibility of constants.

adessy replied:
```ruby
class A
  C1 = 1
  private_constant :C1
  private
  C2 = 2
end

A::C1 # => NameError: private constant A::C1 referenced
A::C2 # => 2
```

koendehondt commented on back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/ref_distributions_controller.rb:
```diff
@@ -41,18 +46,34 @@ def side_fx
+      def infer_ref_distribution_type
+        custom_field = CustomField.find(params[:custom_field_id])
+        if custom_field.key == 'birthyear'
+          'UserCustomFields::Representativeness::BinnedDistribution'
```
Wouldn't it be better to write:
```suggestion
          UserCustomFields::Representativeness::BinnedDistribution.name
```
Same comment on line 64.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/binned_distribution.rb:
```diff
@@ -0,0 +1,89 @@
+      end
+
+      def bin_boundaries
+        distribution['bins'].freeze
```
Given that this method is called multiple times, consider memoizing the return value. Same comment on `counts`.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/categorical_distribution.rb:
```diff
@@ -0,0 +1,94 @@
+      private
+
+      def counts
+        @counts = distribution.values
```
```suggestion
        @counts ||= distribution.values
```

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb:
```diff
@@ -21,71 +22,69 @@
       belongs_to :custom_field
-      has_many :options, through: :custom_field
 
+      validates :type, presence: true, inclusion: { in: ->(_record) { allowed_types } }
```
I do not think that this is necessary. ActiveRecord handles this automatically. It is not possible to create instances of classes in this class hierarchy with a type that does not match any of the class names in the hierarchy. If you like to avoid creation of instances of this class, make it abstract (but I would not worry about it). See citizenlab/back/app/models/application_record.rb for an example.

adessy replied:
Really? I think I tried to use a random string or class name and ActiveRecord didn't complain. He just created an instance of `RefDistribution`.

adessy replied:
Checked again and indeed not necessary.

koendehondt commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb:
```diff
@@ -21,71 +22,69 @@
+        # This method allows pundit to identify the right policy class for
+        # instances of subclasses of this class.
+        def policy_class
+          UserCustomFields::Representativeness::RefDistributionPolicy
```
```suggestion
          ::UserCustomFields::Representativeness::RefDistributionPolicy
```

koendehondt commented on back/engines/commercial/user_custom_fields/app/policies/user_custom_fields/stat_user_policy.rb:
```diff
@@ -6,6 +6,10 @@ def users_by_gender?
     end
 
+    def users_by_age?
+      user&.active? && (user&.admin? || user&.project_moderator?)
```
Instead of repeating this complex condition several times, extract it to a new private method `active_admin_or_project_moderator?`.

adessy committed:
CL-1019 Implement R-Score for birth-year custom field

koendehondt commented on back/engines/commercial/user_custom_fields/app/serializers/user_custom_fields/xlsx/age_stats_serializer.rb:
```diff
@@ -0,0 +1,49 @@
+
+      def age_ranges_column(bins)
+        bins = bins.dup
+        bins[0] = 0 if bins.first.nil?
```
```suggestion
        bins[0] = 0 unless bins.first
```

adessy replied:
Do you think it improves the readability? I like to make it explicit that I expect `nil`, and not `false`. But maybe that's obvious here.

koendehondt replied:
Do you want to make explicit that `bins.first` is `nil`, or do you want to express that there is no first bin? I assumed the latter, because `bins.first` returns `nil` if there is no element in the array. But maybe you wanted to express that the first element in the array holds `nil.`

koendehondt replied:
After looking at the code and the tests again, I understand why you did it. I will close this conversation.

adessy replied:
> But maybe you wanted to express that the first element in the array holds nil.

Yes, exactly!

koendehondt commented on back/engines/commercial/user_custom_fields/app/services/user_custom_fields/age_counter.rb:
```diff
@@ -0,0 +1,54 @@
+    # @param [Array<Numeric,nil>] bins
+    def bin_data(counts, bins)
+      bins = bins.dup
+      bins[0] = 0 if bins.first.nil?
+      bins[-1] = Float::INFINITY if bins.last.nil?
```
```suggestion
      bins[0] = 0 unless bins.first
      bins[-1] = Float::INFINITY unless bins.last
```
Are these boundary conditions described in tests? That is not clear to me.

adessy replied:
There are already quite a few test cases here: `engines/commercial/user_custom_fields/spec/services/user_custom_fields/age_counter_spec.rb:1`. Maybe I could add one with negative ages and check that they are discarded. As for upper bound, I don't see how I could use a test to show it is converted to infinity. How would you go about that?

koendehondt replied:
For the upper bound, isn't it covered by the `realistic_data` case in the test?

koendehondt commented on back/engines/commercial/user_custom_fields/lib/user_custom_fields/core_ext/enumerable.rb:
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
```
Is it a good idea to put an extension to a Ruby module deep down in our code? Shouldn't we put it in an initializer outside the engines?

adessy replied:
I had some hesitation. In the end, I decided to put it in the engine with the idea that we could always move it later to the core if other parts of the code starts using it. But if we were to move it now, I'd still rather keep as a library and `require` it only when needed.

koendehondt commented on back/engines/commercial/user_custom_fields/spec/acceptance/stats_users_spec.rb:
```diff
@@ -630,4 +638,128 @@ def xlsx_worksheet_to_array(worksheet)
+    let(:start_at) { Time.zone.local(2019) }
+    let(:end_at) { Time.zone.local(2020).end_of_year }
+
+    # I made my life easier here, at the expense of test coverage, by using a group
```
Please do not write in the "I" form, because the reader does not know who "I" is.

adessy replied:
...but is the "We" allowed? 😉

koendehondt commented on back/engines/commercial/user_custom_fields/spec/models/user_custom_fields/representativeness/categorical_distribution_spec.rb:
```diff
@@ -2,17 +2,24 @@
   end
 
-  it { is_expected.to belong_to(:custom_field) }
+  # This patch is needed to test counts validations.
```
Doing this may impact other tests, because the patch is not removed. Can't you use https://relishapp.com/rspec/rspec-mocks/docs/configuring-responses/block-implementation ?

adessy replied:
Good catch! I replaced the patch with a helper method.

koendehondt reviewed (changes requested):
Given that my knowledge about the feature is restricted to what I read in the ticket, it is hard for me to assess correctness at the functional level. We have scheduled a session in which you can show me the feature in action. Probably I will ask some review questions there.

adessy committed:
CL-1092 Get the name constant instead of hardcoding the string

Co-authored-by: Koen De Hondt <koen@dehondt.dev>

adessy committed:
CL1019 Ensure constant is from top-level scope

Co-authored-by: Koen De Hondt <koen@dehondt.dev>

adessy committed:
CL-1019 Rework 'reference distribution' shared examples

Rework the 'reference distribution' shared examples to rely on a helper method to set the counts, instead of patching reference distribution classes.

adessy committed:
CL-1092 Merge remote-tracking branch 'origin/CL-1092/binned-counts-reference-distribution' into CL-1092/binned-counts-reference-distribution

adessy committed:
CL-1019 Rephrase comment

adessy committed:
CL-1092 Memoize bin_boundaries

adessy committed:
CL-1092 Memoize counts in reference distribution classes

I had to rework some tests that relied on counts not being memoized.

adessy committed:
CL-1092 Refactor UserCustomFields::StatUserPolicy

adessy committed:
CL-1092 Remove unnecessary validation

koendehondt reviewed (commented):
Responses to your questions on the first pass of the review.

adessy committed:
CL-1092 Add test to age_counter_spec.rb

Check that the default smallest bin boundary is 0.

adessy committed:
Merge pull request #2485 from CitizenLabDotCo/CL-1019/representativeness-score-for-age

adessy committed:
CL-1092 Add comment

adessy committed:
CL-1092 Merge branch 'master' into CL-1092/binned-counts-reference-distribution
